demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs

```diff
@@ -0,0 +1,789 @@
+                .collect::<Vec<_>>(),
+            updated_master_keys: updated_master_keys
+                .iter()
+                .map(|&pk| Some(pk))
+                .collect::<Vec<_>>(),
+            updated_signing_keys_sk_signatures,
+            updated_signing_keys_mk_signatures,
+            updated_master_keys_sk_signatures,
+            updated_master_keys_mk_signatures,
+            wcert_signatures,
+            withdrawal_certificate: Some(withdrawal_certificate),
+            prev_withdrawal_certificate,
+            threshold: Some(threshold),
+            b: b_bool,
+            max_pks,
+            genesis_constant: Some(genesis_constant),
+            genesis_validator_keys_tree_root,
+            is_first_certificate: Some(is_first_certificate),
+            cert_data_hash: Some(cert_data_hash),
+            prev_cert_data_hash: Some(prev_cert_data_hash),
+        }
+    }
+}
+
+impl ConstraintSynthesizer<FieldElement> for NaiveThresholdSignatureWKeyRotation {
+    fn generate_constraints<CS: ConstraintSystemAbstract<FieldElement>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        //Internal checks
+        let log_max_pks = (self.max_pks.next_power_of_two() as u64).trailing_zeros() as usize;
+        assert_eq!(self.max_pks, self.signing_keys.len());
+        assert_eq!(self.max_pks, self.master_keys.len());
+        assert_eq!(self.max_pks, self.updated_signing_keys.len());
+        assert_eq!(self.max_pks, self.updated_master_keys.len());
+        assert_eq!(self.max_pks, self.updated_signing_keys_sk_signatures.len());
+        assert_eq!(self.max_pks, self.updated_signing_keys_mk_signatures.len());
+        assert_eq!(self.max_pks, self.updated_master_keys_sk_signatures.len());
+        assert_eq!(self.max_pks, self.updated_master_keys_mk_signatures.len());
+        assert_eq!(self.max_pks, self.wcert_signatures.len());
+        assert_eq!(log_max_pks + 1, self.b.len());
+        assert!(self.withdrawal_certificate.is_some());
+        assert!(self.withdrawal_certificate.clone().unwrap().custom_fields.len()>0);
+        assert!(self.prev_withdrawal_certificate.is_some());
+        assert!(self.prev_withdrawal_certificate.clone().unwrap().custom_fields.len()>0);
+
+        // 1 *************** Verify certificates **************
+        // Alloc field elements for wcert
+        let withdrawal_certificate_g = WithdrawalCertificateDataGadget::alloc(
+            cs.ns(|| "alloc wcert data"), 
+            || self.withdrawal_certificate.clone().ok_or(SynthesisError::AssignmentMissing)
+        )?;
+
+        let threshold_g = FieldElementGadget::alloc(cs.ns(|| "alloc threshold"), || {
+            self.threshold.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+
+
+        let mut skeys_g = Vec::with_capacity(self.max_pks);
+        let mut mkeys_g = Vec::with_capacity(self.max_pks);
+        let mut validator_mktree_leaves_g = Vec::with_capacity(self.max_pks);
```

Comment:
doesn't look like its the prev validator leaves since its using the signing and master keys.

Reply:
Well this will be checked against the previous certificate, so we can call it "prev" as the other ones are called "upd", or we can call this "actual", as you wish